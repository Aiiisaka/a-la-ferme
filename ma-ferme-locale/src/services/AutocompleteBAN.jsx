
import React from "react";

import { debounce } from "lodash"
import { Autocomplete, Box, Grid, TextField, Typography } from "@mui/material";
import axios from "axios";
import LocationOnIcon from '@mui/icons-material/LocationOn';

export default function AutocompleteBAN({ maxResults, setLocation }) {

    const [value, setValue] = React.useState(null);
    const [inputValue, setInputValue] = React.useState('');
    const [options, setOptions] = React.useState([]);

    const fetch = React.useMemo( // memoize the fetching of the data to avoid unnecessary requests
        () =>
            debounce((request, callback) => {
                axios.get('http://api-adresse.data.gouv.fr/search/?q=' + request + '&limit=' + maxResults)
                    .then(response => {
                        callback(response.data.features);
                    }
                    )
                    .catch(error => {
                        console.log(error);
                    });
            }, 500),
        [],
    );

    React.useEffect(() => {
        let active = true;

        if (inputValue === '') {
            setOptions(value ? [value] : []);
            return undefined;
        }

        fetch(inputValue, (results) => {
            if(active) {
                let newOptions = [];

                if (value) {
                    newOptions = [value];
                }

                if(results) {
                    newOptions = [...newOptions, ...results];
                }
                
                setOptions(newOptions);
            }
        });

        return () => {
            active = false;
        };

    }, [value, inputValue, fetch]);

    return (
        <Autocomplete
            id="combo-box-demo"
            options={options}
            filterOptions={(x) => x}
            autoComplete
            includeInputInList
            filterSelectedOptions
            value={value}
            getOptionLabel={option => option.properties.name + ', ' + option.properties.postcode + ' ' + option.properties.city}
            isOptionEqualToValue={(option, value) => option.properties.label === value.properties.label}
            onChange={(event, newValue) => {
                setOptions(newValue ? [newValue, ...options] : options);
                setValue(newValue);
                setLocation(newValue.geometry.coordinates[0], newValue.geometry.coordinates[1]);
            }}
            onInputChange={(event, newInputValue) => {
                setInputValue(newInputValue);
            }}
            renderInput={params => (<TextField {...params} label="Rechercher une adresse" variant="outlined" />)}
            renderOption={(props, option) => {
                return (
                    <li {...props}>
                        <Grid container alignItems="center">
                            <Grid item>
                                <Box
                                    component={LocationOnIcon}
                                    sx={{ color: 'text.secondary', mr: 2 }}
                                />
                            </Grid>
                            <Grid item xs>
                                    <span key={option.properties.id}>
                                        {option.properties.name}
                                    </span>

                                <Typography variant="body2" color="text.secondary">
                                    {option.properties.postcode + ' ' + option.properties.city}
                                </Typography>
                            </Grid>
                        </Grid>
                    </li>
                );
            }}

        />
    );
}



