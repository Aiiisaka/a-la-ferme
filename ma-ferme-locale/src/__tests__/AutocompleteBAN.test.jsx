import { cleanup, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AutocompleteBAN from '../components/Farm/Positions/AutocompleteBAN';

describe('AutocompleteBAN components', () => {

    afterEach(cleanup);

    it('renders correctly', () => {
        const tree = renderer.create(<AutocompleteBAN />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('It\'s different after click', () => {
        const { getByText } = render(<AutocompleteBAN />);
        getByText.find('input').simulate('click');
        expect(getByText).not.toMatchSnapshot();
    });

});

describe('AutocompleteBAN components', () => {

    afterEach(cleanup);

    //try to write an address in the input field and check if the autocomplete is working
    it('renders correctly', () => {
        const { getByTestId } = render(<AutocompleteBAN />);
        const input = document.querySelector('input');
        input.value = '1 rue de la paix';
        expect(input.value).toBe('1 rue de la paix');
    });

});
