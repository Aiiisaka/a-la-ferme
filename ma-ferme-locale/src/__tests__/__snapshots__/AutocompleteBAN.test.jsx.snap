// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteBAN components renders correctly 1`] = `
<div
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor="combo-box-demo"
      id="combo-box-demo-label"
    >
      Rechercher une adresse
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="both"
        aria-expanded={false}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="combo-box-demo"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        readOnly={false}
        required={false}
        role="combobox"
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-1ftyaf0"
        >
          <span>
            Rechercher une adresse
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
